### PR TITLE
docs(core): add Typedoc comments

### PR DESCRIPTION
**Pull request by baumstern**

### CODE REVIEW

File: core/ts/utils/utils.ts [hunk 1]
```diff
@@ -32,26 +58,46 @@ const genProcessVkSig = (
...
+ * @returns Returns a signature for querying if a verifying key with the given parameters is already registered in the contract.
+ */
 const genTallyVkSig = (_stateTreeDepth: number, _intStateTreeDepth: number, _voteOptionTreeDepth: number): bigint => {
   return (
     (BigInt(_stateTreeDepth) << BigInt(128)) + (BigInt(_intStateTreeDepth) << BigInt(64)) + BigInt(_voteOptionTreeDepth)
   );
 };
 
+/**
+ * This function generates a signature for querying if a verifying key of the SubsidyCalculation circuit with the given parameters is already registered in the contract.
```
Comment (ctrlc03) [inline core/ts/utils/utils.ts:75]: Same as above for Tally votes sig

File: core/ts/utils/utils.ts [hunk 2]
```diff
@@ -62,6 +108,11 @@ const packProcessMessageSmallVals = (
   return packedVals;
 };
 
+/**
+ * This function unpacks the following values from a single bigint:
```
Comment (ctrlc03) [inline core/ts/utils/utils.ts:112]: The following values are?

File: core/ts/utils/utils.ts [hunk 3]
```diff
@@ -81,13 +132,25 @@ const unpackProcessMessageSmallVals = (packedVals: bigint) => {
...
+ */
+const packTallyVotesSmallVals = (batchStartIndex: number, batchSize: number, numSignUps: number): bigint => {
   // Note: the << operator has lower precedence than +
   const packedVals = BigInt(batchStartIndex) / BigInt(batchSize) + (BigInt(numSignUps) << BigInt(50));
 
   return packedVals;
 };
 
+/**
+ * This function unpacks the following values from a single bigint:
```
Comment (ctrlc03) [inline core/ts/utils/utils.ts:150]: The following values are?

File: core/ts/Poll.ts [hunk 4]
```diff
@@ -395,8 +409,9 @@ class Poll implements IPoll {
    * poll has concluded.
    * @param _pollId The ID of the poll associated with the messages to
    *        process
+   * @returns stringified circuit inputs
    */
-  public processMessages = (_pollId: number): any => {
+  public processMessages = (_pollId: number): string[] => {
```
Comment (ctrlc03) [inline core/ts/Poll.ts:414]: why is this an array of strings? thought it'd be an object

File: core/ts/Poll.ts [hunk 5]
```diff
@@ -857,9 +903,10 @@ class Poll implements IPoll {
   };
 
   /**
-   * Tally a batch of Ballots and update this.results
+   * This method tallies a ballots and updates the tally results.
+   * @returns the circuit inputs for the TallyVotes circuit.
    */
-  public tallyVotes = () => {
+  public tallyVotes = (): unknown => {
```
Comment (ctrlc03) [inline core/ts/Poll.ts:909]: Why unknown here, where for instance on processMessages you added a specific type?

File: core/ts/Poll.ts [hunk 6]
```diff
@@ -709,16 +727,28 @@ class Poll implements IPoll {
...
+   * @returns The tally of votes
+   */
+  public hasUntalliedBallots = (): boolean => {
     const batchSize = this.batchSizes.tallyBatchSize;
     return this.numBatchesTallied * batchSize < this.ballots.length;
   };
 
-  public hasUnfinishedSubsidyCalculation = () => {
+  /**
+   * This method checks if there are any unfinished subsidy calculations.
```
[the remark this answers is not in the document]
Comment (baumstern) [reply core/ts/Poll.ts:740]: ok, I'm going to apply a line length limit of 80 here

File: core/ts/Poll.ts [hunk 7]
```diff
@@ -785,18 +815,21 @@ class Poll implements IPoll {
     return circuitInputs as bigint[];
   };
 
-  public increaseSubsidyIndex = () => {
+  private increaseSubsidyIndex = () => {
```
[the remark this answers is not in the document]
Comment (baumstern) [reply core/ts/Poll.ts:818]: fair enough

File: core/ts/Poll.ts (hunk 4 above)
Comment (baumstern) [reply core/ts/Poll.ts:414]: type madness of `stringifyBigInts()` 🫠

**Review by ctrlc03 — CHANGES_REQUESTED** [review]
Left a couple more comments - thanks for the changes so far

File: core/ts/Poll.ts [hunk 8]
```diff
@@ -28,30 +28,30 @@ const blankStateLeafHash = blankStateLeaf.hash();
...
-  hasUnprocessedMessages(): boolean;
+  topupMessage(_message: Message): void;
+
+  // These methods are used to generate circuit inputs
   processMessages(_pollId: number): any;
-  genProcessMessagesCircuitInputsPartial(_index: number): any;
+  tallyVotes(): any;
+
+  // These methods are helper functions
+  mergeAllMessages(): void;
```
[the remark this answers is not in the document]
Comment (baumstern) [reply core/ts/Poll.ts:40]: definetely not!

File: core/ts/MaciState.ts [hunk 9]
```diff
@@ -5,15 +5,12 @@ import { Poll } from "./Poll";
...
 
-/*
- * File Overview:
- * - IMaciState: Interface detailing `MaciState`'s public API.
- * - MaciState: Core class implementing the above interface.
+/**
+ * Represents the public API of the MaciState class.
  */
-
 interface IMaciState {
```
Comment (0xmad) [inline core/ts/MaciState.ts:11]: Let's use consistent new lines. I'd suggest to remove all the newlines between interface props unless you'd like to add description for props.

File: core/ts/Poll.ts (hunk 8 above)
Comment (0xmad) [inline core/ts/Poll.ts:40]: The same here with new line consistency.

File: core/ts/Poll.ts [hunk 10]
```diff
@@ -370,6 +384,7 @@ class Poll implements IPoll {
    * poll has concluded.
    * @param _pollId The ID of the poll associated with the messages to
    *        process
+   * @returns stringified circuit inputs
    */
   public processMessages = (_pollId: number): any => {
```
[the remark this answers is not in the document]
Comment (0xmad) [reply core/ts/Poll.ts:389]: I have a better option :)

```
type CircuitInputs  = Record<string, string | bigint | bigint[] | string[]>;
```

File: core/ts/__tests__/MaciState.test.ts [hunk 11]
```diff
@@ -275,7 +275,7 @@ describe("MaciState", function () {
     it("should tally ballots correctly", () => {
       // Start with results = [0...0]
       let total = BigInt(0);
-      for (const v of poll.results) {
+      for (const v of maciState.polls[pollId].tallyResult) {
         total = total + v;
       }
```
Comment (0xmad) [inline core/ts/__tests__/MaciState.test.ts:280]:
```suggestion
      const total = calculateTotal(maciState.polls[pollId].tallyResult);
```

File: core/ts/__tests__/MaciState.test.ts [hunk 12]
```diff
@@ -141,7 +141,7 @@ describe("MaciState", function () {
 
     it("Tally ballots", () => {
       let total = BigInt(0);
-      for (const v of poll.results) {
+      for (const v of maciState.polls[pollId].tallyResult) {
         total = BigInt(Number(total) + Number(v));
       }
```
Comment (0xmad) [inline core/ts/__tests__/MaciState.test.ts:146]:
```suggestion
     const calculateTotal = (tallyResult: TallyResult): bigint => tallyResult.reduce((acc, v) => acc + v, BigInt(0));

      const total = calculateTotal(maciState.polls[pollId].tallyResult);
```

File: core/ts/__tests__/MaciState.test.ts [hunk 13]
```diff
@@ -151,7 +151,7 @@ describe("MaciState", function () {
       poll.tallyVotes();
 
       total = BigInt(0);
-      for (const v of poll.results) {
+      for (const v of maciState.polls[pollId].tallyResult) {
         total = BigInt(Number(total) + Number(v));
       }
```
Comment (0xmad) [inline core/ts/__tests__/MaciState.test.ts:156]:
```suggestion
      const total = calculateTotal(maciState.polls[pollId].tallyResult);
```

**Review by 0xmad — CHANGES_REQUESTED** [review]
@baumstern thanks, just some changes requested

File: core/ts/__tests__/MaciState.test.ts (hunk 11 above)
Comment (baumstern) [reply core/ts/__tests__/MaciState.test.ts:280]: Given `calculateTotal()` going to be used only this code block for testing purpose, I'll use following code block instead:
```
// Check if all elements in tallyResult are 0
      const allZero = maciState.polls[pollId].tallyResult.every(v => v === BigInt(0));
      expect(allZero).to.eq(true);
```

File: core/package.json [hunk 14]
```diff
@@ -12,6 +12,7 @@
     "test": "ts-mocha --exit ts/__tests__/*.test.ts"
   },
   "dependencies": {
+    "maci-core": "^1.1.2",
```
Comment (0xmad) [inline core/package.json:15]: Why do you need `maci-core` here?

Comment (baumstern) [reply core/package.json]: ah it's a typo

**Review by 0xmad — DISMISSED** [review]
@baumstern thanks!

File: core/ts/utils/utils.ts [hunk 15]
```diff
@@ -32,26 +65,53 @@ const genProcessVkSig = (
...
+ * @returns Returns a signature for querying if a verifying key with
+ * the given parameters is already registered in the contract.
+ */
 const genTallyVkSig = (_stateTreeDepth: number, _intStateTreeDepth: number, _voteOptionTreeDepth: number): bigint => {
   return (
     (BigInt(_stateTreeDepth) << BigInt(128)) + (BigInt(_intStateTreeDepth) << BigInt(64)) + BigInt(_voteOptionTreeDepth)
   );
 };
 
+/**
```
Comment (ctrlc03) [inline core/ts/utils/utils.ts:84]: there seems to be some repetition of comments here

File: core/ts/utils/utils.ts [hunk 16]
```diff
@@ -62,6 +122,11 @@ const packProcessMessageSmallVals = (
   return packedVals;
 };
 
+/**
+ * This function packs it's parameters into a single bigint.
```
Comment (ctrlc03) [inline core/ts/utils/utils.ts:126]: it's the opposite

File: core/ts/utils/utils.ts [hunk 17]
```diff
@@ -81,13 +146,25 @@ const unpackProcessMessageSmallVals = (packedVals: bigint) => {
...
+ */
+const packTallyVotesSmallVals = (batchStartIndex: number, batchSize: number, numSignUps: number): bigint => {
   // Note: the << operator has lower precedence than +
   const packedVals = BigInt(batchStartIndex) / BigInt(batchSize) + (BigInt(numSignUps) << BigInt(50));
 
   return packedVals;
 };
 
+/**
+ * This function packs it's parameters into a single bigint.
```
Comment (ctrlc03) [inline core/ts/utils/utils.ts:164]: It's the opposite

File: core/ts/utils/utils.ts [hunk 18]
```diff
@@ -100,6 +177,13 @@ const unpackTallyVotesSmallVals = (packedVals: bigint) => {
...
+ * @param row - The row.
+ * @param col - The column.
+ * @param numSignUps - The number of sign ups.
+ * @returns Returns a single bigint that contains the packed values.
+ */
 const packSubsidySmallVals = (row: number, col: number, numSignUps: number) => {
```
Comment (ctrlc03) [inline core/ts/utils/utils.ts:187]: do we have an unpack function for subsidy values?

**Review by ctrlc03 — CHANGES_REQUESTED** [review]
Left couple more comments! thanks

File: core/ts/utils/utils.ts (hunk 18 above)
Comment (baumstern) [reply core/ts/utils/utils.ts:186]: there's no need to have it

**Comment by baumstern** [comment]
@ctrlc03 thanks, please take another look

**Review by ctrlc03 — APPROVED** [review]
Great stuff, LGTM!

**Pull request merged**